device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -485,13 +500,35 @@ public void registerDeviceClients(Iterable<DeviceClient> deviceClients, long tim
                 log.info("Registering device {} to multiplexing client", configBeingRegistered.getDeviceId());
             }
 
-            this.deviceIO.registerMultiplexedDeviceClient(deviceClientConfigsToRegister, timeoutMilliseconds);
+            Set<String> deviceIdsThatFailedToRegister = new HashSet<>();
+            MultiplexingClientDeviceRegistrationAuthenticationException registrationAuthenticationException = null;
+            try
+            {
+                this.deviceIO.registerMultiplexedDeviceClient(deviceClientConfigsToRegister, timeoutMilliseconds);
+            }
+            catch (MultiplexingClientDeviceRegistrationAuthenticationException e)
+            {
+                // If registration failed, 1 or more clients should not be considered registered in this layer's state.
+                // Save the exception so it can be rethrown once the local state has been updated to match the actual state
+                // of the multiplexed connection.
+                deviceIdsThatFailedToRegister = e.getRegistrationExceptions().keySet();
+                registrationAuthenticationException = e;
+            }
 
-            for (DeviceClient successfullyRegisteredClient : deviceClients)
+            for (DeviceClient clientsThatAttemptedToRegister : deviceClients)
             {
                 // Only update the local state map once the register call has succeeded
-                String registeredDeviceId = successfullyRegisteredClient.getConfig().getDeviceId();
-                this.multiplexedDeviceClients.put(registeredDeviceId, successfullyRegisteredClient);
+                String deviceIdThatAttemptedToRegister = clientsThatAttemptedToRegister.getConfig().getDeviceId();
+
+                if (!deviceIdsThatFailedToRegister.contains(deviceIdThatAttemptedToRegister))
+                {
+                    this.multiplexedDeviceClients.put(deviceIdThatAttemptedToRegister, clientsThatAttemptedToRegister);
```

Comment:
can we not do this on line 508 instead of iterating through the list again?

Reply:
if I am reading this right though, the try block (506 to 508) is not inside a loop, is it?

Reply:
It should to be inside a bigger loop right?

Reply:
There was a bigger loop earlier that I could leverage, so I removed this loop in the happy path case. For the case where 1 or more devices fail to register, we still need to compare the list of devices that failed against the list of devices that didn't